From b4b32ae6d5267126a582de1203c37d4be6ce06e0 Mon Sep 17 00:00:00 2001
From: lzc256 <i@lzc256.com>
Date: Mon, 1 Jun 2026 21:11:22 +0800
Subject: [PATCH] Patch: Android anet - fix network interface access without
 root

- Replace net.Interfaces() with anet.Interfaces()
- Replace iface.Addrs() with anet.InterfaceAddrsByInterface()
- Keep net import for net.FlagUp, net.IPNet, net.Addr
---
 client/firewall/uspfilter/localip.go    | 5 +++--
 client/internal/stdnet/discover_pion.go | 7 +++----
 client/system/network_addr.go           | 6 ++++--
 3 files changed, 10 insertions(+), 8 deletions(-)

diff --git a/client/firewall/uspfilter/localip.go b/client/firewall/uspfilter/localip.go
index b35be56..9595be2 100644
--- a/client/firewall/uspfilter/localip.go
+++ b/client/firewall/uspfilter/localip.go
@@ -3,6 +3,7 @@ package uspfilter
 import (
 	"fmt"
 	"net"
+	"github.com/wlynxg/anet"
 	"net/netip"
 	"sync/atomic"

@@ -30,7 +31,7 @@ func newLocalIPManager() *localIPManager {
 }

 func processInterface(iface net.Interface, ips map[netip.Addr]struct{}, addresses *[]netip.Addr) {
-	addrs, err := iface.Addrs()
+	addrs, err := anet.InterfaceAddrsByInterface(&iface)
 	if err != nil {
 		log.Debugf("get addresses for interface %s failed: %v", iface.Name, err)
 		return
@@ -84,7 +85,7 @@ func (m *localIPManager) UpdateLocalIPs(iface common.IFaceMapper) (err error) {
 		}
 	}

-	interfaces, err := net.Interfaces()
+	interfaces, err := anet.Interfaces()
 	if err != nil {
 		log.Warnf("failed to get interfaces: %v", err)
 	} else {
diff --git a/client/internal/stdnet/discover_pion.go b/client/internal/stdnet/discover_pion.go
index 83b4a1d..5e16448 100644
--- a/client/internal/stdnet/discover_pion.go
+++ b/client/internal/stdnet/discover_pion.go
@@ -1,9 +1,8 @@
 package stdnet

 import (
-	"net"
-
 	"github.com/pion/transport/v3"
+	"github.com/wlynxg/anet"
 )

 type pionDiscover struct {
@@ -12,7 +11,7 @@ type pionDiscover struct {
 func (d pionDiscover) iFaces() ([]*transport.Interface, error) {
 	ifs := []*transport.Interface{}

-	oifs, err := net.Interfaces()
+	oifs, err := anet.Interfaces()
 	if err != nil {
 		return nil, err
 	}
@@ -20,7 +19,7 @@ func (d pionDiscover) iFaces() ([]*transport.Interface, error) {
 	for _, oif := range oifs {
 		ifc := transport.NewInterface(oif)

-		addrs, err := oif.Addrs()
+		addrs, err := anet.InterfaceAddrsByInterface(&oif)
 		if err != nil {
 			return nil, err
 		}
diff --git a/client/system/network_addr.go b/client/system/network_addr.go
index 5423cf8..e80bd68 100644
--- a/client/system/network_addr.go
+++ b/client/system/network_addr.go
@@ -5,10 +5,12 @@ package system
 import (
 	"net"
 	"net/netip"
+
+	"github.com/wlynxg/anet"
 )

 func networkAddresses() ([]NetworkAddress, error) {
-	interfaces, err := net.Interfaces()
+	interfaces, err := anet.Interfaces()
 	if err != nil {
 		return nil, err
 	}
@@ -21,7 +23,7 @@ func networkAddresses() ([]NetworkAddress, error) {
 		if iface.HardwareAddr.String() == "" {
 			continue
 		}
-		addrs, err := iface.Addrs()
+		addrs, err := anet.InterfaceAddrsByInterface(&iface)
 		if err != nil {
 			continue
 		}
--
2.53.0
